src/drivers/intel/cavs/dmic.c

```diff
@@ -1444,8 +1444,14 @@ static int dmic_probe(struct dai *dai)
 {
 	struct dmic_pdata *dmic;
 
-	trace_dmic("pro");
+	if (dai_get_drvdata(dai))
+		return 0; /* already probed */
 
+	trace_dmic("pro");
+#if defined(CONFIG_APOLLOLAKE)
+	/* Disable dynamic clock gating for dmic before touching any register */
+	shim_write(SHIM_CLKCTL, shim_read(SHIM_CLKCTL) | SHIM_CLKCTL_DMICFDCGB);
```

Comment:
I want to align us more with Linux here since the Linux APIs are well understood/documented. This should be a call to 
```
/* Disable dynamic clock gating for dmic before touching any register */
pm_runtime_get_sync(DMIC_CLK, dmic->id); 
```
where DMIC_CLK is defined in platform and dmic->id is the particular DMIC index. The clock gating would be enabled again with a call to 
```
pm_runtime_put(DMIC_CLK, dmic->id); 
```